README.md
# node-solr-lambda

[![Conventional Commits](https://img.shields.io/badge/Conventional%20Commits-1.0.0-yellow.svg)](https://conventionalcommits.org)

## Features

- functional-style API with jsdoc type annotations
- plain javascipt in commonjs format - no additional build step required
- ES6+ including async/await, lambdas, spread operator...
- typechecking support in vscode (without compiling typescript code)
- no runtime dependencies other than `axios`
- as close as possible to the Solr Json-based REST API
- functions for working with documents, fields, field types, cores

## Example

```js
const { prepareSolrClient } = require("node-solr-lambda");
const solr = prepareSolrClient("mycore");

async function myfun() {
  const result = await solr.query({ query: "label:something" });
  console.log(result);

  await solr.addField({ name: "price", type: "plong", multiValued: false });
  await solr.addField({
    name: "category",
    type: "string",
    multiValued: true,
    docValues: true
  });

  const item1 = { id: "item1", price: 123, category: ["cpu", "ram"] };
  const item2 = { id: "item2", price: 456, category: ["cpu", "usb"] };
  await solr.add(item1); // works with single object
  await solr.add([item1, item2]); // works with object[]
  await solr.commit()

  await solr.query({
    query: "*:*",
    facet: {
      top_5_categories: {
        terms: {
          field: "category",
          limit: 5,
          facet: {
            avg_price: "avg(price)"
          }
        }
      }
    }
  });
}
```

## How to test

We provide a docker-compose with a solr instance for testing.
The testing solr instance needs to run first:
```sh
# run this in a separate terminal and shutdown using CTRL+C
yarn test:prepare
```

Now you can run the test suite:
```sh
yarn test
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [3.0.5](https://github.com/vsimko/node-solr-lambda/compare/v3.0.4...v3.0.5) (2023-06-27)

### [3.0.4](https://github.com/vsimko/node-solr-lambda/compare/v3.0.3...v3.0.4) (2020-05-12)

### [3.0.3](https://github.com/vsimko/node-solr-lambda/compare/v3.0.1...v3.0.3) (2020-05-12)


### Bug Fixes

* link to types.d.ts mentioned in jsdocs ([c732f02](https://github.com/vsimko/node-solr-lambda/commit/c732f02bdd16a5be414658b4d92bbc46187af3b7))

<a name="3.0.1"></a>
## [3.0.1](https://github.com/vsimko/node-solr-lambda/compare/v3.0.0...v3.0.1) (2018-12-09)


### Bug Fixes

* update docs, move solr.d.ts to types.d.ts ([af72e46](https://github.com/vsimko/node-solr-lambda/commit/af72e46))



<a name="3.0.0"></a>
# [3.0.0](https://github.com/vsimko/node-solr-lambda/compare/v2.2.0...v3.0.0) (2018-12-08)


### Bug Fixes

* convert 404 and 400 http errors to exceptions ([6ca25ed](https://github.com/vsimko/node-solr-lambda/commit/6ca25ed))


### Features

* split api between client and core admin and simplify internal design ([37b0788](https://github.com/vsimko/node-solr-lambda/commit/37b0788))


### BREAKING CHANGES

* - config:
  - removed parameters debug, apiPrefix
  - urlConfig moved up one level
- printing of debugging messages removed
- prepareSolrClient accepts first parameter `core` and optinoal parameter `userConfig`
- added function `commit`
- added function `solrListFields`
- moved function `ping` to `prepareCoreAdmin`
- added function `solrDeleteCore` to `prepareCoreAdmin`
- functions `deleteField` and `deleteFieldType` now accepts a single `name` parameter rather than object



<a name="2.2.0"></a>
# [2.2.0](https://github.com/vsimko/node-solr-lambda/compare/v2.1.0...v2.2.0) (2018-12-05)


### Bug Fixes

* function `mergeConfig` now uses `deepCopy` through JSON serdes. ([c240ad3](https://github.com/vsimko/node-solr-lambda/commit/c240ad3))


### Features

* add `mergedConfig` function ([9ca2fe6](https://github.com/vsimko/node-solr-lambda/commit/9ca2fe6))



<a name="2.1.0"></a>
# [2.1.0](https://github.com/vsimko/node-solr-lambda/compare/v2.0.2...v2.1.0) (2018-12-05)


### Features

* add solr.config function and and configAutoEnableFields ([ade5e8a](https://github.com/vsimko/node-solr-lambda/commit/ade5e8a))



<a name="2.0.2"></a>
## [2.0.2](https://github.com/vsimko/node-solr-lambda/compare/v2.0.0...v2.0.2) (2018-12-04)



<a name="2.0.1"></a>
## [2.0.1](https://github.com/vsimko/node-solr-lambda/compare/v2.0.0...v2.0.1) (2018-12-04)



<a name="2.0.0"></a>
# [2.0.0](https://github.com/vsimko/node-solr-lambda/compare/v1.1.0...v2.0.0) (2018-12-04)


### Bug Fixes

* add explicit typedef imports instead of declare global ([c388079](https://github.com/vsimko/node-solr-lambda/commit/c388079))


### Features

* add mergeConfig exported function used also in prepareSolrClient ([051ecd0](https://github.com/vsimko/node-solr-lambda/commit/051ecd0))


### BREAKING CHANGES

* the prepareSolrClient function now accepts
user config which is merged with the default config (overrides its
fields). The userConfig can also be empty.



<a name="1.1.0"></a>
# [1.1.0](https://github.com/vsimko/node-solr-lambda/compare/v1.0.0...v1.1.0) (2018-12-04)


### Bug Fixes

* printing curl command in debug mode should use quoted data ([752c1cc](https://github.com/vsimko/node-solr-lambda/commit/752c1cc))


### Features

* implemented `delete` function for deleting document by id or query ([e1166e4](https://github.com/vsimko/node-solr-lambda/commit/e1166e4))



<a name="1.0.0"></a>
# [1.0.0](https://github.com/vsimko/node-solr-lambda/compare/v0.2.2...v1.0.0) (2018-12-04)


### Bug Fixes

* export defaultConfig ([773c947](https://github.com/vsimko/node-solr-lambda/commit/773c947))
* problem with `query` param accidentally renamed ([b1fbe8c](https://github.com/vsimko/node-solr-lambda/commit/b1fbe8c))
* remove solrGet and use solrPost everywhere, update typedefs ([6837233](https://github.com/vsimko/node-solr-lambda/commit/6837233))


### Features

* add debugging of GET requests ([d609b65](https://github.com/vsimko/node-solr-lambda/commit/d609b65))
* add type annotations, new solr functions ([253f335](https://github.com/vsimko/node-solr-lambda/commit/253f335))
* add typings mainly for nicer vscode assistance ([9d3604e](https://github.com/vsimko/node-solr-lambda/commit/9d3604e))
* facet function to get solr facets using JSON api ([c7cfa78](https://github.com/vsimko/node-solr-lambda/commit/c7cfa78))


### BREAKING CHANGES

* - `select` function replaced by `query`
- `facet` function removed because we now have query+typedefs
- we now use the Solr JSON API in `SolrQuery` type with additional parameters inside `SolrQuery.params`.
- the config has now nested section `urlConfig` which is passed directly to `url.format` before calling `axios.post`.
* From now on,  the `core` parameter needs
to be passed through the `config` parameter. This simplifies
the API by removing special cases.
The provided typings provide better user experience when developing
in an IDE such as vscode.



<a name="0.2.2"></a>
## [0.2.2](https://github.com/vsimko/node-solr-lambda/compare/v0.2.1...v0.2.2) (2018-07-31)



<a name="0.2.1"></a>
## [0.2.1](https://github.com/vsimko/node-solr-lambda/compare/v0.1.1...v0.2.1) (2018-07-31)



<a name="0.1.1"></a>
## 0.1.1 (2018-06-20)


### Bug Fixes

* some typos ([c469209](https://github.com/vsimko/node-solr-lambda/commit/c469209))


test/solr.test.js
// @ts-check
const { prepareSolrClient, prepareCoreAdmin } = require("../src/index")
const { expect, use } = require("chai")
const cap = require("chai-as-promised")
use(cap)

/** @typedef {import("../src/types").SolrConfig} SolrConfig */

/** @type {SolrConfig} */
const testConfing = { port: 18983 } // see docker-compose.yml

describe("#prepareCoreAdmin", async () => {
  const admin = prepareCoreAdmin(testConfing)

  it("should ping core", () => {
    admin.ping("testcore")
    expect(admin.ping("non-existent-core")).to.be.rejectedWith("Not Found")
  })

  it("should delete core", () => {
    expect(admin.solrDeleteCore("non-existent-core")).to.be.rejectedWith(
      "Cannot unload non-existent core"
    )
  })
})

describe("#prepareSolrClient", () => {
  const solr = prepareSolrClient("testcore", testConfing)

  it("should query documents", async () => {
    await solr.query({ query: "*:*" })
  })

  it("should delete documents", async () => {
    await solr.delete({ id: "123" })
  })

  it("should add documents and commit", async () => {
    await solr.add({ id: "123", hello: "world" })
    await solr.commit()
  })

  it("should provide other functions", () => {
    expect(solr).to.haveOwnProperty("deleteField")
    expect(solr).to.haveOwnProperty("mergedConfig")
    expect(solr).to.haveOwnProperty("deleteFieldType")
    expect(solr).to.haveOwnProperty("addField")
    expect(solr).to.haveOwnProperty("addFieldType")
    expect(solr).to.haveOwnProperty("replaceField")
    expect(solr).to.haveOwnProperty("config")
    expect(solr).to.haveOwnProperty("solrListFields")
    expect(solr).to.haveOwnProperty("configAutoEnableFields")
  })
})


src/utils.js
// @ts-check
/**
 * Merges `b` configuration structure to `a` by creating
 * a deep copy of `a`.
 * - Note: The configuration structure `a` must be serializable to JSON.
 * - This function is pure
 *
 * @template T
 * @param {T} a
 * @param {T} b
 */
const mergeConfig = (a, b) => mergeConfigImpure(deepCopy(a), b)

/**
 * Creates a deep copy of an object using JSON seriazlie/deserialize.
 * This function is pure
 * @template T
 * @param {T} x Must be serializable to JSON!
 * @return {T}
 */
const deepCopy = x => JSON.parse(JSON.stringify(x))

/**
 * Merges `b` to `a` by altering `a`.
 * @template T
 * @param {T} target
 * @param {T} source
 */
function mergeConfigImpure (target, source) {
  for (const k in source) {
    const maybeObject = target[k]
    if (maybeObject != null && maybeObject.constructor === Object) {
      mergeConfigImpure(maybeObject, source[k]) // recurse on objects
    } else {
      target[k] = deepCopy(source[k]) // assign scalar value
    }
  }
  return target
}

/**
 * @template T
 * @param {T | T[]} x
 * @return {T[]}
 */
const ensureArray = x => x instanceof Array ? x : [x]

module.exports = {
  mergeConfig,
  mergeConfigImpure,
  ensureArray
}


src/solr.d.ts
import { AxiosError } from "axios"
import { ClientRequest } from "http"
import { UrlObject } from "url"

// TODO: similarity https://lucene.apache.org/solr/guide/7_5/other-schema-elements.html#similarity
// TODO: unique key https://lucene.apache.org/solr/guide/7_5/other-schema-elements.html#unique-key
// TODO: dynamic fields https://lucene.apache.org/solr/guide/7_5/dynamic-fields.html
// TODO: copying fields https://lucene.apache.org/solr/guide/7_5/copying-fields.html
// https://lucene.apache.org/solr/guide/7_5/v2-api.html

/**
 * Field types that are available in Solr (except deprecated fields).
 * @see https://lucene.apache.org/solr/guide/7_5/field-types-included-with-solr.html#field-types-included-with-solr
 */
type FieldTypesIncludedInSolr =
  | "solr.BinaryField"
  | "solr.BoolField"
  | "solr.CollationField"
  | "solr.CurrencyFieldType"
  | "solr.DateRangeField"
  | "solr.DatePointField"
  | "solr.DoublePointField"
  | "solr.ExternalFileField"
  | "solr.EnumFieldType"
  | "solr.FloatPointField"
  | "solr.ICUCollationField"
  | "solr.IntPointField"
  | "solr.LatLonPointSpatialField"
  | "solr.LongPointField"
  | "solr.PointType"
  | "solr.PreAnalyzedField"
  | "solr.RandomSortField"
  | "solr.SpatialRecursivePrefixTreeFieldType"
  | "solr.StrField"
  | "solr.StrField"
  | "solr.TextField"
  | "solr.UUIDField"

/**
 * Field definitions must have a `name` and `class`.
 * @see https://lucene.apache.org/solr/guide/7_5/defining-fields.html#field-properties
 */
interface FieldGeneralProperties {
  /**
   * The name of the field. Field names should consist of alphanumeric or underscore characters
   * only and not start with a digit. This is not currently strictly enforced, but other field
   * names will not have first class support from all components and back compatibility is not
   * guaranteed. Names with both leading and trailing underscores (e.g., _version_) are reserved.
   * Every field must have a name.
   */
  name: string

  /**
   * The name of the fieldType for this field. This will be found in the name attribute on
   * the fieldType definition. Every field must have a type.
   */
  type: string

  /**
   * A default value that will be added automatically to any document that does not have a value
   * in this field when it is indexed. If this property is not specified, there is no default.
   */
  default?: string
}

/**
 */
interface FieldTypeGeneralProperties {
  /**
   * The name of the fieldType. This value gets used in field definitions,
   * in the "type" attribute. It is strongly recommended that names consist
   * of alphanumeric or underscore characters only and not start with a digit.
   * This is not currently strictly enforced.
   */
  name: string

  /**
   * The class name that gets used to store and index the data for this type.
   * Note that you may prefix included class names with "solr." and Solr will
   * automatically figure out which packages to search for the class - so
   * `solr.TextField` will work.
   *
   * If you are using a third-party class, you will probably need to have a fully
   * qualified class name. The fully qualified equivalent for `solr.TextField`
   * is `org.apache.solr.schema.TextField`.
   */
  class: FieldTypesIncludedInSolr // | string

  /**
   * For multivalued fields, specifies a distance between multiple values,
   * which prevents spurious phrase matches.
   */
  positionIncrementGap?: number

  /**
   * For text fields. If true, Solr automatically generates phrase queries
   * for adjacent terms. If false, terms must be enclosed in double-quotes
   * to be treated as phrases.
   */
  autoGeneratePhraseQueries?: boolean

  /**
   * Query used to combine scores of overlapping query terms (i.e., synonyms).
   * Consider a search for "blue tee" with query-time synonyms `tshirt,tee`.
   * - Use `as_same_term` (default) to blend terms, i.e., `SynonymQuery(tshirt,tee)`
   *   where each term will be treated as equally important. The value `as_same_term`
   *   is appropriate when terms are true synonyms (television, tv).
   * - Use `pick_best` to select the most significant synonym when scoring `Dismax(tee,tshirt)`.
   * - Use `as_distinct_terms` to bias scoring towards the most significant synonym (`pants OR slacks`).
   * - Use `pick_best` or `as_distinct_terms` when synonyms are expanding to hyponyms
   *   (`q=jeans w/ jeans=>jeans,pants`) and you want exact to come before parent and
   *   sibling concepts.
   * @see http://opensourceconnections.com/blog/2017/11/21/solr-synonyms-mea-culpa/
   */
  synonymQueryStyle?: "as_same_term" | "pick_best" | "as_distinct_terms"

  /**
   * For text fields,applicable when querying with
   * [`sow=false`](https://lucene.apache.org/solr/guide/7_5/the-standard-query-parser.html#standard-query-parser-parameters)
   * (which is the default for the sow parameter).
   * - Use `true`, the default, for field types with query analyzers including graph-aware filters,
   *   e.g., [Synonym Graph Filter](https://lucene.apache.org/solr/guide/7_5/filter-descriptions.html#synonym-graph-filter)
   *   and [Word Delimiter Graph Filter](https://lucene.apache.org/solr/guide/7_5/filter-descriptions.html#word-delimiter-graph-filter).
   * - Use `false` for field types with query analyzers including filters that
   *   can match docs when some tokens are missing,
   *   e.g., [Shingle Filter](https://lucene.apache.org/solr/guide/7_5/filter-descriptions.html#shingle-filter).
   */
  enableGraphQueries?: boolean

  /**
   * Defines a custom `DocValuesFormat` to use for fields of this type.
   * This requires that a schema-aware codec, such as the `SchemaCodecFactory`
   * has been configured in `solrconfig.xml`.
   */
  docValuesFormat?: string

  /**
   * Defines a custom `PostingsFormat` to use for fields of this type.
   * This requires that a schema-aware codec, such as the `SchemaCodecFactory`
   * has been configured in `solrconfig.xml`.
   *
   * **Note:**
   * Lucene index back-compatibility is only supported for the default codec.
   * If you choose to customize the `postingsFormat` or `docValuesFormat` in your
   * `schema.xml`, upgrading to a future version of Solr may require you to either
   * switch back to the default codec and optimize your index to rewrite it into
   * the default codec before upgrading, or re-build your entire index from scratch
   * after upgrading.
   */
  postingsFormat?: string
}

/**
 * Fields can have many of the same properties as field types.
 * @see https://lucene.apache.org/solr/guide/7_5/field-type-definitions-and-properties.html#field-default-properties
 */
interface FieldTypeDefaultProperties {
  /** If true, the value of the field can be used in queries to retrieve matching documents. */
  indexed?: boolean

  /** If true, the actual value of the field can be retrieved by queries. */
  stored?: boolean

  /**
   * If true, the value of the field will be put in a column-oriented
   * [DocValues](https://lucene.apache.org/solr/guide/7_5/docvalues.html#docvalues) structure.
   */
  docValues?: boolean

  /** Control the placement of documents when a sort field is not present. */
  sortMissingFirst?: boolean

  /** Control the placement of documents when a sort field is not present. */
  sortMissingLast?: boolean

  /** If true, indicates that a single document might contain multiple values for this field type. */
  multiValued?: boolean

  /**
   * If true, omits the norms associated with this field (this disables length normalization
   * for the field, and saves some memory). **Defaults to true for all primitive (non-analyzed)
   * field types, such as int, float, data, bool, and string.** Only full-text fields or fields need norms.
   */
  omitNorms?: boolean

  /**
   * If true, omits term frequency, positions, and payloads from postings for this field.
   * This can be a performance boost for fields that don’t require that information.
   * It also reduces the storage space required for the index. Queries that rely on position
   * that are issued on a field with this option will silently fail to find documents.
   * **This property defaults to true for all field types that are not text fields.**
   */
  omitTermFreqAndPositions?: boolean

  /**
   * Similar to `omitTermFreqAndPositions` but preserves term frequency information.
   */
  omitPositions?: boolean

  /**
   * These options instruct Solr to maintain full term vectors for each document,
   * optionally including position, offset and payload information for each term
   * occurrence in those vectors. These can be used to accelerate highlighting and
   * other ancillary functionality, but impose a substantial cost in terms of index size.
   * They are not necessary for typical uses of Solr.
   */
  termVectors?: boolean
  termPositions?: boolean
  termOffsets?: boolean
  termPayloads?: boolean

  /**
   * Instructs Solr to reject any attempts to add a document which does not have a value
   * for this field. This property defaults to false.
   */
  required?: boolean

  /**
   * If the field has docValues enabled, setting this to true would allow the field
   * to be returned as if it were a stored field (even if it has `stored=false`) when
   * matching `“*”` in an fl parameter.
   */
  useDocValuesAsStored?: boolean

  /**
   * Large fields are always lazy loaded and will only take up space in the document
   * cache if the actual value is < 512KB. This option requires `stored="true"` and
   * `multiValued="false"`. It’s intended for fields that might have very large values
   * so that they don’t get cached in memory.
   */
  large?: boolean
}

type SolrFragmentWithId = {
  id: string
  [key: string]: string | string[]
}

// TODO: SolrDataValue
type SolrDataValue = any
// type SolrDataValue =
//   | string
//   | string[]
//   | SolrFragmentWithId
//   | SolrFragmentWithId[]

interface TermsFacet {
  buckets: { val: string; count: number }[]
}

interface FacetFunctionResult {
  [facetName: string]: number
}

export interface SolrData {
  [key: string]: SolrDataValue
}

/**
 * Fields are defined in the fields element of `schema.xml`.
 * Once you have the **field types** set up, defining the fields themselves is simple.
 *
 * @see https://lucene.apache.org/solr/guide/7_5/defining-fields.html
 * @see https://lucene.apache.org/solr/guide/7_5/defining-fields.html#optional-field-type-override-properties
 */
export type FieldProperties = FieldGeneralProperties &
  FieldTypeDefaultProperties

/**
 * A field type defines the analysis that will occur on a field when documents are indexed
 * or queries are sent to the index.
 * A field type definition can include four types of information:
 * - The `name` of the field type (**mandatory**).
 * - An implementation `class` name (**mandatory**).
 * - If the field type is `TextField`, a description of the field analysis for the field type.
 * - Field type properties - depending on the implementation class,
 *   some properties **may be mandatory**.
 *
 * The field type `class` determines most of the behavior of a field type, but optional properties
 * can also be defined.
 *
 * The properties that can be specified for a given field type fall into three major categories:
 * - Properties specific to the field type’s class.
 * - General Properties {@link FieldTypeGeneralProperties} that Solr supports for any field type.
 * - Field Default Properties {@link FieldTypeDefaultProperties} that can be specified on the field
 *   type that will be inherited by fields that use this type instead of the default behavior.
 */
export type FieldTypeProperties = FieldTypeGeneralProperties &
  FieldTypeDefaultProperties

export interface SolrDocument {
  id: string
  _childDocuments_?: SolrFragmentWithId | SolrFragmentWithId[]
  [key: string]: SolrDataValue | SolrData
}

/**
 * This type definition contains just the most important parts.
 */
interface SolrResponseHeader {
  params?: object
  status: number
  QTime: number
}

/**
 * This type definition contains just the most important parts.
 */
export interface SolrResponse {
  status?: string
  facets?: {
    count?: number
  } & {
    [facetName: string]: (TermsFacet | FacetFunctionResult)[] // TODO: to be completed
  }
  response?: {
    docs: SolrDocument[]
    numFound: number
    start: number
  }
  responseHeader: SolrResponseHeader
}

/**
 * This type definition contains just the most important parts.
 */
export interface SolrException {
  config
  message: string
  request: ClientRequest
  response: {
    config
    data: {
      error: {
        code: number
        details: {
          errorMessages: string[]
          [key: string]: object
        }[]
        metadata: string[]
      }
      responseHeader: SolrResponseHeader
    }
    headers
    request: ClientRequest
    status: number
    statusText: string
  }
  stack: string
}

export interface QueryRequest {
  query?
  filter?
  start?
  limit?
  sort?
  facet?
  params?: {
    hl?: "on" | "off"
    "hl.simple.pre"?: string
    "hl.simple.post"?: string
    "hl.fl"?: string
    indent?: "off" | "on"
  }
}

export interface SolrConfig {
  urlConfig?: UrlObject & {
    query?: {
      overwrite?: boolean
      commitWithin?: number
      wt?: "json" | "xml" | "python" | "ruby" | "php" | "csv"
      [key: string]: any
    }
  }
  debug?: boolean
  core?: string
  apiPrefix?: string
}

export interface DeleteRequest {
  id?: string
  query?: any
}

export type ConfigRequest = {
  // Commands for Common Properties:
  // https://lucene.apache.org/solr/guide/7_5/config-api.html#commands-for-common-properties

  "set-property"?: { [property: string]: any }
  "unset-property"?: string

  // Basic Commands for Components
  // https://lucene.apache.org/solr/guide/7_5/config-api.html#basic-commands-for-components

  "add-requesthandler"?: any
  "update-requesthandler"?: any
  "delete-requesthandler"?: any
  "add-searchcomponent"?: any
  "update-searchcomponent"?: any
  "delete-searchcomponent"?: any
  "add-initparams"?: any
  "update-initparams"?: any
  "delete-initparams"?: any
  "add-queryresponsewriter"?: any
  "update-queryresponsewriter"?: any
  "delete-queryresponsewriter"?: any

  // Advanced Commands for Components:
  // https://lucene.apache.org/solr/guide/7_5/config-api.html#advanced-commands-for-components

  "add-queryparser"?: any
  "update-queryparser"?: any
  "delete-queryparser"?: any
  "add-valuesourceparser"?: any
  "update-valuesourceparser"?: any
  "delete-valuesourceparser"?: any
  "add-transformer"?: any
  "update-transformer"?: any
  "delete-transformer"?: any
  "add-updateprocessor"?: any
  "update-updateprocessor"?: any
  "delete-updateprocessor"?: any
  "add-queryconverter"?: any
  "update-queryconverter"?: any
  "delete-queryconverter"?: any
  "add-listener"?: any
  "update-listener"?: any
  "delete-listener"?: any
  "add-runtimelib"?: any
  "update-runtimelib"?: any
  "delete-runtimelib"?: any

  // Commands for User-Defined Properties:
  // https://lucene.apache.org/solr/guide/7_5/config-api.html#commands-for-user-defined-properties
  "set-user-property"?: {
    "update.autoCreateFields"?: TrueOrFalseString // boolean does not work, perhaps a bug in solr
    [variableName: string]: any
  }
  "unset-user-property"?: string
}

type TrueOrFalseString = "true" | "false"
type OnOrOffString = "on" | "off"

src/index.js
// @ts-check
/**
 * Support for type checking and intellisense in vscode:
 * @typedef {import("./types").SolrConfig} SolrConfig
 * @typedef {import("./types").ConfigRequest} ConfigRequest
 * @typedef {import("./types").SolrException} SolrException
 * @typedef {import("./types").SolrDocument} SolrDocument
 * @typedef {import("./types").SolrResponse} SolrResponse
 * @typedef {import("./types").DeleteRequest} DeleteRequest
 * @typedef {import("./types").FieldProperties} FieldProperties
 * @typedef {import("./types").FieldTypeProperties} FieldTypeProperties
 * @typedef {import("./types").QueryRequest} QueryRequest
 * @typedef {{name:string, type?:string, class?:string, enumsConfig?:string, enumName?:string, [key:string]:any}} FieldDef
 * @typedef {import("axios").AxiosResponse} AxiosResponse
 */

const url = require("url")
const { default: axios } = require("axios")
const { mergeConfig, ensureArray } = require("./utils")

/** @type {SolrConfig} */
const defaultConfig = {
  hostname: "localhost",
  port: 8983,
  protocol: "http",
  query: {
    commitWithin: 500,
    overwrite: true
  }
}

/** @type {(axiosResp:AxiosResponse) => SolrResponse} */
const convertSolrResponse = axiosResp => axiosResp.data

// converting error response from solr
/** @param {SolrException} reason */
const convertSolrError = reason => {
  switch (reason.response.status) {
    case 400:
      throw Error(reason.response.data.error.msg)
    case 404:
      throw Error(`${reason.message}: ${reason.response.statusText}`)
    default:
      throw reason
  }
}

/**
 * @param {SolrConfig} userConfig This is merged with the defaultConfig
 * @param {string} core
 */
const prepareSolrClient = (core, userConfig = {}) => {
  console.assert(core, "Missing 'core' parameter")

  const config = mergeConfig(defaultConfig, userConfig)

  /** @type {(pathname:string) => (data:any) => Promise<SolrResponse>} */
  const httpPostReq = pathname => data =>
    axios
      .post(url.format({ ...config, pathname }), data)
      .then(convertSolrResponse)
      .catch(convertSolrError)

  /** @type {(pathname:string) => (query:any) => Promise<SolrResponse>} */
  const httpGetReq = pathname => query =>
    axios
      .get(url.format({ ...mergeConfig(config, { query }), pathname }))
      .then(convertSolrResponse)
      .catch(convertSolrError)

  /** @type {(op:string) => (fieldDef:FieldDef) => Promise<SolrResponse>} */
  const solrSchemaReq = op => fieldDef =>
    httpPostReq(`/api/cores/${core}/schema`)({ [op]: fieldDef })

  /** @type {(data:ConfigRequest) => Promise<SolrResponse> } */
  const configReq = httpPostReq(`/api/cores/${core}/config`)

  // now creating the API
  return {
    mergedConfig: () => config,

    commit: () => httpGetReq(`/solr/${core}/update`)({ commit: true }),

    /** @type {(data:QueryRequest) => Promise<SolrResponse>} */
    query: httpPostReq(`/api/cores/${core}/query`),

    /** @param {SolrDocument | SolrDocument[]} data */
    add: data => httpPostReq(`/api/cores/${core}/update`)(ensureArray(data)),

    /** @param {DeleteRequest} deleteQuery */
    delete: deleteQuery =>
      httpPostReq(`/api/cores/${core}/update`)({ delete: deleteQuery }),

    /** @param {string} name */
    deleteField: name => solrSchemaReq("delete-field")({ name }),

    /** @param {string} name */
    deleteFieldType: name => solrSchemaReq("delete-field-type")({ name }),

    /** @type {(data:FieldProperties) => Promise<SolrResponse>} */
    addField: solrSchemaReq("add-field"),

    /** @type {(data:FieldTypeProperties) => Promise<SolrResponse>} */
    addFieldType: solrSchemaReq("add-field-type"),

    /** @type {(data:FieldTypeProperties) => Promise<SolrResponse>} */
    replaceField: solrSchemaReq("replace-field"),

    config: configReq,

    /** @type {() => Promise<string[]>} */
    solrListFields: () =>
      httpGetReq(`/solr/${core}/schema/fields`)({}).then(data =>
        data.fields.map(field => field.name)
      ),

    /** @param {boolean} enable */
    configAutoEnableFields: enable =>
      configReq({
        "set-user-property": {
          "update.autoCreateFields": enable ? "true" : "false"
        }
      })
  }
}

/** @param {SolrConfig} userConfig */
const prepareCoreAdmin = userConfig => {
  const config = mergeConfig(defaultConfig, userConfig)
  return {
    /** @type {(core:string) => Promise<SolrResponse>} */
    ping: core =>
      axios
        .get(url.format({ ...config, pathname: `/solr/${core}/admin/ping` }))
        .then(convertSolrResponse)
        .catch(convertSolrError),

    /** @type {(core:string) => Promise<SolrResponse>} */
    solrDeleteCore: core =>
      axios
        .get(
          url.format({
            ...mergeConfig(config, {
              query: {
                core,
                action: "UNLOAD",
                deleteIndex: true,
                deleteDataDir: true,
                deleteInstanceDir: true
              }
            }),
            pathname: "/solr/admin/cores"
          })
        )
        .then(convertSolrResponse)
        .catch(convertSolrError)
  }
}

module.exports = {
  prepareSolrClient,
  prepareCoreAdmin,
  defaultConfig,
  mergeConfig
}
